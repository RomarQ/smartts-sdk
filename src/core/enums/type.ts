export enum TypeAtom {
    record = 'record',
    variant = 'variant',
    chain_id = 'chain_id',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    option = 'option',
    map = 'map',
    big_map = 'bigmap',
    signature = 'signature',
    operation = 'operation',
    never = 'never',
    tuple = 'tuple',
    lambda = 'lambda',
    ticket = 'ticket',
    sapling_state = 'sapling_state',
    contract = 'contract',
    sapling_transaction = 'sapling_transaction',
}

export default TypeAtom;
