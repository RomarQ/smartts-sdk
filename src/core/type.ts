import { IToString } from '../typings';

export class BaseType {
    private _ = 'Type';
}

export class SimpleType extends BaseType {
    constructor(public name: string) {
        super();
    }

    toString() {
        return `"${this.name}"`;
    }
}

export class T_TList extends BaseType {
    constructor(public innerType: IToString) {
        super();
    }

    toString() {
        return `(list ${this.innerType})`;
    }
}

export const TList = (innerType: BaseType) => new T_TList(innerType);

export const TNat = new SimpleType('nat');
export const TInt = new SimpleType('int');
export const TMutez = new SimpleType('mutez');
export const TString = new SimpleType('string');
