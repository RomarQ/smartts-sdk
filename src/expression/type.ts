import type { IExpression } from '../typings/expression';
import type { IType } from '../typings/type';
import { Expression } from '../core/expression';
import ExpressionAtom from '../core/enums/expression';
import { LineInfo } from '../misc/utils';
import { proxy } from '../misc/proxy';

/**
 * Annotate an expression with type.
 *
 * ```typescript
 * AsType(GetVariable("some_variable"), TNat());
 * ```
 *
 * @param expression
 * @param type
 * @param {LineInfo} line Source code line information (Used in error messages)
 *
 * @category | Type Handling
 *
 * @returns {IExpression} An expression
 */
export const AsType = (expression: IExpression, type: IType, line = new LineInfo()): IExpression<any> =>
    proxy(new Expression(ExpressionAtom.type_annotation, expression, type, line), Expression.proxyHandler);
