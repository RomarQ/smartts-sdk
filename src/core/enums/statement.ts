export enum StatementAtom {
    verify = 'verify',
    ifBlock = 'ifBlock',
    elseBlock = 'elseBlock',
    forGroup = 'forGroup',
    whileBlock = 'whileBlock',
    set = 'set',
    defineLocal = 'defineLocal',
    result = 'result',
    failwith = 'failwith',
    match = 'match',
    match_cases = 'match_cases',
    set_type = 'set_type',
}

export default StatementAtom;
