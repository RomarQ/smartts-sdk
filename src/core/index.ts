import type { IToString, IToType } from '../typings';
import type { ILiteral } from '../typings/literal';
import type { IType } from '../typings/type';

import { Expression } from './expression';
import { Mutez } from './literal';
import Utils, { LineInfo } from '../misc/utils';
import { TUnit } from './type';

export class Flag {
    args: string[] = [];

    constructor(public flag: string, ...args: string[]) {
        this.args = args;
    }

    toString() {
        return `(${this.flag}${(' ' + this.args.join(' ')).trim()})`;
    }
}

interface EntryPointOptions {
    mock?: boolean;
    lazy?: boolean;
    lazyAndCodeless?: boolean;
}
export class EntryPoint {
    options = {
        mock: false,
        lazy: false,
        lazyAndCodeless: false,
    };
    inType?: IType;
    statements: IToString[] = [];

    constructor(public name: string, public line = new LineInfo()) {}

    config(options?: EntryPointOptions) {
        if (options?.mock) {
            this.options.mock = options.mock;
        }
        if (options?.lazy) {
            this.options.lazy = options.lazy;
        }
        if (options?.lazyAndCodeless) {
            this.options.lazyAndCodeless = options.lazyAndCodeless;
        }
        return this;
    }

    inputType(type: IType) {
        this.inType = type;
        return this;
    }

    code(callback: (arg: IToString) => IToString[]) {
        this.statements = callback(new Expression('params', new LineInfo()));
        return this;
    }

    toString() {
        const notMock = Utils.capitalizeBoolean(!this.options.mock);
        const isLazy = Utils.capitalizeBoolean(this.options.lazy);
        const isLazyAndCodeless = Utils.capitalizeBoolean(this.options.lazyAndCodeless);
        const hasParams = Utils.capitalizeBoolean(!!this.inType);
        return `(${this.name} ${notMock} ${isLazy} ${isLazyAndCodeless} ${hasParams} ${
            this.line
        } (${this.statements.join(' ')}))`;
    }
}

export class Contract {
    options: {
        initialBalance: ILiteral;
        flags: Flag[];
    } = {
        initialBalance: Mutez(0),
        flags: [],
    };
    public initialStorage: IToString & IToType;
    public entries: EntryPoint[];

    constructor(
        args: {
            initialStorage: IToString & IToType;
            entries: EntryPoint[];
        },
        public line = new LineInfo(),
    ) {
        this.initialStorage = args.initialStorage;
        this.entries = args.entries;
    }

    config(options?: { initialBalance?: ILiteral; flags?: Flag[] }) {
        if (options?.flags) {
            this.options.flags = options.flags;
        }
        if (options?.initialBalance) {
            this.options.initialBalance = options.initialBalance;
        }
        return this;
    }

    toString() {
        return `
        (
            template_id (static_id 0 ${this.line})
            storage ${this.initialStorage.toString()}
            storage_type (${this.initialStorage.toType()})
            messages (${this.entries.map((entry) => entry.toString()).join(' ')})
            flags (${this.options.flags.map((flag) => flag.toString())})
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance ${this.options.initialBalance}
        )
        `;
    }
}

export { default as Command } from './command';
export { default as Expression } from './expression';
export { default as Type } from './type';
export { default as Literal } from './literal';
export * from './blockchain_operations';
