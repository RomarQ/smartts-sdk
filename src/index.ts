// Load polyfills needed by SmartML
import './compiler/polyfills';

export * from './core';
export * from './type';
export * from './expression';
export * from './statement';
