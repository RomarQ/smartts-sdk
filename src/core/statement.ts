import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import { LineInfo } from '../misc/utils';
import StatementAtom from './enums/statement';
import { Expression } from './expression';
import ExpressionAtom from './enums/expression';
import { Proxied, proxy } from '../misc/proxy';
import { Add, GetVariable, Iterator, LessThanOrEqual } from '../expression';
import { NewVariable, SetValue } from '../statement';

export class Statement implements IStatement {
    args;

    constructor(public name: string, ...args: (IExpression | IType | LineInfo | string | ILiteral)[]) {
        this.args = args || [];
    }

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}

export class IfStatment implements IStatement {
    #condition: IExpression;
    #thenStatements: IStatement[];
    #elseStatements: IStatement[];
    #line: LineInfo;

    constructor(
        condition: IExpression,
        line = new LineInfo(),
        thenStatements?: IStatement[],
        elseStatements?: IStatement[],
    ) {
        this.#condition = condition;
        this.#thenStatements = thenStatements || [];
        this.#elseStatements = elseStatements || [];
        this.#line = line;
    }

    public Then(statements: IStatement[]) {
        this.#thenStatements = statements;
        return this;
    }

    public Else(statements: IStatement[]) {
        this.#elseStatements = statements;
        return this;
    }

    private ifBlock() {
        return `(${StatementAtom.ifBlock} ${this.#condition} (${this.#thenStatements.join(' ')}) ${this.#line})`;
    }

    private elseBlock() {
        return `(${StatementAtom.elseBlock} (${this.#elseStatements.join(' ')}))`;
    }

    [Symbol.toPrimitive]() {
        let expression = this.ifBlock();
        if (this.#elseStatements.length) {
            expression = `${expression} ${this.elseBlock()}`;
        }
        return expression;
    }
}

export class VariantMatchStatement implements IStatement {
    static idCounter = 0;
    private cases: Record<string, IStatement[]> = {};

    static get nextID() {
        return ++VariantMatchStatement.idCounter;
    }

    constructor(private variant: IExpression, private argumentName: string, private line = new LineInfo()) {}

    public setArgumentName(argumentName: string): this {
        this.argumentName = argumentName;
        return this;
    }

    public Case(branch: string, buildStatements: (arg: Proxied<IExpression>) => IStatement[]): this {
        const variantArgument = this.variantArgument(`${this.argumentName}_${branch}`, this.line);
        this.cases[branch] = buildStatements(variantArgument);
        return this;
    }

    private caseArgument(argumentName: string, line: LineInfo) {
        return new Expression(ExpressionAtom.cases_arg, `"${argumentName}"`, line);
    }

    private variantArgument(argumentName: string, line: LineInfo) {
        return proxy(new Expression(ExpressionAtom.variant_arg, `"${argumentName}"`, line), Expression.proxyHandler);
    }

    [Symbol.toPrimitive]() {
        const caseArgument = this.caseArgument(this.argumentName, this.line);
        const matchCases = Object.keys(this.cases).map((branch) => {
            return new Statement(
                StatementAtom.match,
                caseArgument,
                branch,
                `"${this.argumentName}_${branch}"`,
                `(${this.cases[branch].join(' ')})`,
                this.line,
            );
        });
        return `(${StatementAtom.match_cases} ${this.variant} "${this.argumentName}" (${matchCases.join(' ')}) ${
            this.line
        })`;
    }
}

export class WhileStatement implements IStatement {
    constructor(private condition: IExpression, private statements: IStatement[] = [], private line = new LineInfo()) {}

    public Do(statements: IStatement[]) {
        this.statements = statements;
        return this;
    }

    [Symbol.toPrimitive]() {
        return `(${StatementAtom.whileBlock} ${this.condition} (${this.statements.join(' ')}) ${this.line})`;
    }
}

export class ForStatement implements IStatement {
    static idCounter = 0;

    static get nextID() {
        return ++VariantMatchStatement.idCounter;
    }

    constructor(
        private from: IExpression,
        private to: IExpression,
        private increment: IExpression,
        private statements: IStatement[] = [],
        private iteratorName: string,
        private line = new LineInfo(),
    ) {}

    public setIteratorName(iteratorName: string): this {
        this.iteratorName = iteratorName;
        return this;
    }

    public Do(callback: (iterator: Proxied<IExpression>) => IStatement[], line = new LineInfo()) {
        const iterator = GetVariable(this.iteratorName, line);
        this.statements = callback(iterator);
        return this;
    }

    [Symbol.toPrimitive]() {
        const variable = NewVariable(this.iteratorName, this.from);
        const condition = LessThanOrEqual(GetVariable(this.iteratorName), this.to);
        const stmts = [
            ...this.statements,
            SetValue(GetVariable(this.iteratorName), Add(GetVariable(this.iteratorName), this.increment)),
        ];
        return `${variable} (${StatementAtom.whileBlock} ${condition} (${stmts.join(' ')}) ${this.line})`;
    }
}

export class ForEachStatement implements IStatement {
    static idCounter = 0;

    static get nextID() {
        return ++VariantMatchStatement.idCounter;
    }

    constructor(
        private list: IExpression,
        private statements: IStatement[],
        private iteratorName: string,
        private line = new LineInfo(),
    ) {}

    public setIteratorName(iteratorName: string): this {
        this.iteratorName = iteratorName;
        return this;
    }

    public Do(callback: (iterator: Proxied<IExpression>) => IStatement[], line = new LineInfo()) {
        const iterator = Iterator(this.iteratorName, line);
        this.statements = callback(iterator);
        return this;
    }

    [Symbol.toPrimitive]() {
        return `(${StatementAtom.forGroup} "${this.iteratorName}" ${this.list} (${this.statements.join(' ')}) ${
            this.line
        })`;
    }
}
