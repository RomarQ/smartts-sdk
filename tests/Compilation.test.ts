import { TList, TNat, TOption } from '../src/type';
import {
    Address,
    List,
    Nat,
    String,
    None,
    Record,
    Some,
    Unit,
    ContractStorage,
    Equal,
    GreaterThanOrEqual,
    GetVariable,
    GetSender,
    Lambda,
    Mutez,
} from '../src/expression';
import { Contract, EntryPoint, Flag, OnChainView } from '../src/core';
import { FailWith, If, NewVariable, Require, Return, SetValue } from '../src/statement';
import { verifyContractCompilationOutput, verifyLambdaCompilationOutput } from './util';

describe('Compile Lambdas', () => {
    it('A Lambda that returns the argument', () => {
        const lambda = Lambda().code((arg) => [
            If(Equal(arg, String('TEST')))
                .Then([Return(arg)])
                .Else([FailWith(arg)]),
        ]);

        verifyLambdaCompilationOutput(lambda);
    });
    it(' A Lambda that returns "YES" if the argument is greater than or equal to Nat(10), returns "NO" otherwise', () => {
        const lambda = Lambda().code((arg) => [
            If(GreaterThanOrEqual(arg, Nat(1)))
                .Then([Return(String('YES'))])
                .Else([Return(String('NO'))]),
        ]);

        verifyLambdaCompilationOutput(lambda);
    });
});

describe('Compile Contract', () => {
    it('Contract 1', () => {
        const contract = new Contract()
            .setStorageType(TList(TNat()))
            .setStorage(List([]))
            .setConfig({
                initialBalance: Mutez(100),
                flags: [new Flag('erase-comments')],
            })
            .addEntrypoint(
                new EntryPoint('ep1')
                    .config({ lazy: false })
                    .inputType(TList(TNat()))
                    .code((arg) => [
                        // Define a variable named "some_address"
                        NewVariable('some_address', Address('tz1')),
                        // Require sender to be equal to variable "some_address", otherwise fail with "Not Admin!"
                        Require(Equal(GetVariable('some_address'), GetSender()), String('Not Admin!')),
                        // Replace the storage value with entry point argument
                        SetValue(ContractStorage(), arg),
                    ]),
            )
            .addView(new OnChainView('view').code((argument) => [Return(argument)]));
        verifyContractCompilationOutput(contract);
    });

    it('Storage (Unit)', () => {
        const contract = new Contract().setStorage(Unit());

        verifyContractCompilationOutput(contract);
    });

    it('Storage (Some)', () => {
        const contract = new Contract().setStorage(Some(Nat(1)));

        verifyContractCompilationOutput(contract);
    });
    it('Storage (None)', () => {
        const contract = new Contract().setStorageType(TOption(TNat())).setStorage(None());

        verifyContractCompilationOutput(contract);
    });

    it('Storage (Record)', () => {
        const contract = new Contract().setStorage(
            Record({
                testField1: Nat(1),
                testField2: List([String('Hello World')]),
            }),
        );

        verifyContractCompilationOutput(contract);
    });
});
