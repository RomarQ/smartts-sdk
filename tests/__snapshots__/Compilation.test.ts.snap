// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compile Contract Contract 1 1`] = `
"
        (
            template_id (static_id 0 (\\"Compilation.test.ts\\" 52))
            storage (list (\\"Compilation.test.ts\\" 54) )
            storage_type ((list \\"nat\\"))
            messages ((ep1 True False False True (\\"Compilation.test.ts\\" 60) ((set_type (params (\\"Compilation.test.ts\\" 60)) (list \\"nat\\") (\\"Compilation.test.ts\\" 74)) (defineLocal \\"some_address\\" (literal (address tz1) (\\"Compilation.test.ts\\" 63)) True (\\"Compilation.test.ts\\" 63)) (verify (eq (getLocal \\"some_address\\" (\\"Compilation.test.ts\\" 63)) (sender) (\\"Compilation.test.ts\\" 63)) (literal (string \\"Not Admin!\\") (\\"Compilation.test.ts\\" 63)) (\\"Compilation.test.ts\\" 63)) (set (data) (params (\\"Compilation.test.ts\\" 60)) (\\"Compilation.test.ts\\" 63)))))
            flags ((erase-comments))
            privates ()
            views ((onchain view True (\\"Compilation.test.ts\\" 72) False \\"\\" ((set_type (params (\\"Compilation.test.ts\\" 72)) \\"bool\\" (\\"Compilation.test.ts\\" 74)) (result (params (\\"Compilation.test.ts\\" 72)) (\\"Compilation.test.ts\\" 72)))))
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 100) (\\"Compilation.test.ts\\" 56))
        )
        "
`;

exports[`Compile Contract Contract 1 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"list\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"nat\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"list\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"nat\\"
                    }
                ],
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CAR\\"
                },
                {
                    \\"prim\\": \\"PUSH\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"address\\"
                        },
                        {
                            \\"string\\": \\"tz1\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"SENDER\\"
                },
                {
                    \\"prim\\": \\"COMPARE\\"
                },
                {
                    \\"prim\\": \\"EQ\\"
                },
                {
                    \\"prim\\": \\"IF\\",
                    \\"args\\": [
                        [],
                        [
                            {
                                \\"prim\\": \\"PUSH\\",
                                \\"args\\": [
                                    {
                                        \\"prim\\": \\"string\\"
                                    },
                                    {
                                        \\"string\\": \\"Not Admin!\\"
                                    }
                                ]
                            },
                            {
                                \\"prim\\": \\"FAILWITH\\"
                            }
                        ]
                    ]
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    },
    {
        \\"prim\\": \\"view\\",
        \\"args\\": [
            {
                \\"string\\": \\"view\\"
            },
            {
                \\"prim\\": \\"bool\\"
            },
            {
                \\"prim\\": \\"bool\\"
            },
            [
                {
                    \\"prim\\": \\"CAR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Compile Contract Contract 2 1`] = `
"
        (
            template_id (static_id 0 (\\"Compilation.test.ts\\" 78))
            storage (literal (nat 0) (\\"Compilation.test.ts\\" 80))
            storage_type (())
            messages ((entry_point_1 True False False True (\\"Compilation.test.ts\\" 83) ((set_type (params (\\"Compilation.test.ts\\" 83)) \\"bool\\" (\\"Compilation.test.ts\\" 104)) (set (data) (openVariant (view \\"negate\\" (self_address) (params (\\"Compilation.test.ts\\" 83)) (unknown 0) (\\"Compilation.test.ts\\" 87)) \\"Some\\" (variant \\"Some\\" (literal (string \\"Could not call view\\") (\\"Compilation.test.ts\\" 87)) (\\"Compilation.test.ts\\" 87)) (\\"Compilation.test.ts\\" 87)) (\\"Compilation.test.ts\\" 87)))))
            flags ()
            privates ()
            views ((onchain negate True (\\"Compilation.test.ts\\" 96) False \\"\\" ((set_type (params (\\"Compilation.test.ts\\" 96)) \\"bool\\" (\\"Compilation.test.ts\\" 104)) (ifBlock (params (\\"Compilation.test.ts\\" 96)) ((result (literal (nat 1) (\\"Compilation.test.ts\\" 96)) (\\"Compilation.test.ts\\" 96))) (\\"Compilation.test.ts\\" 96)) (elseBlock ((result (literal (nat 0) (\\"Compilation.test.ts\\" 96)) (\\"Compilation.test.ts\\" 96)))))))
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"Compilation.test.ts\\" 78))
        )
        "
`;

exports[`Compile Contract Contract 2 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"nat\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"bool\\",
                \\"annots\\": [
                    \\"%entry_point_1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CAR\\"
                },
                {
                    \\"prim\\": \\"SELF_ADDRESS\\"
                },
                {
                    \\"prim\\": \\"SWAP\\"
                },
                {
                    \\"prim\\": \\"VIEW\\",
                    \\"args\\": [
                        {
                            \\"string\\": \\"negate\\"
                        },
                        {
                            \\"prim\\": \\"nat\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"IF_NONE\\",
                    \\"args\\": [
                        [
                            {
                                \\"prim\\": \\"PUSH\\",
                                \\"args\\": [
                                    {
                                        \\"prim\\": \\"option\\",
                                        \\"args\\": [
                                            {
                                                \\"prim\\": \\"string\\"
                                            }
                                        ]
                                    },
                                    {
                                        \\"prim\\": \\"Some\\",
                                        \\"args\\": [
                                            {
                                                \\"string\\": \\"Could not call view\\"
                                            }
                                        ]
                                    }
                                ]
                            },
                            {
                                \\"prim\\": \\"FAILWITH\\"
                            }
                        ],
                        []
                    ]
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    },
    {
        \\"prim\\": \\"view\\",
        \\"args\\": [
            {
                \\"string\\": \\"negate\\"
            },
            {
                \\"prim\\": \\"bool\\"
            },
            {
                \\"prim\\": \\"nat\\"
            },
            [
                {
                    \\"prim\\": \\"CAR\\"
                },
                {
                    \\"prim\\": \\"IF\\",
                    \\"args\\": [
                        [
                            {
                                \\"prim\\": \\"PUSH\\",
                                \\"args\\": [
                                    {
                                        \\"prim\\": \\"nat\\"
                                    },
                                    {
                                        \\"int\\": \\"1\\"
                                    }
                                ]
                            }
                        ],
                        [
                            {
                                \\"prim\\": \\"PUSH\\",
                                \\"args\\": [
                                    {
                                        \\"prim\\": \\"nat\\"
                                    },
                                    {
                                        \\"int\\": \\"0\\"
                                    }
                                ]
                            }
                        ]
                    ]
                }
            ]
        ]
    }
]"
`;

exports[`Compile Contract Storage (None) 1`] = `
"
        (
            template_id (static_id 0 (\\"Compilation.test.ts\\" 119))
            storage (variant \\"None\\" (unit) (\\"Compilation.test.ts\\" 119))
            storage_type ((option \\"nat\\"))
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"Compilation.test.ts\\" 119))
        )
        "
`;

exports[`Compile Contract Storage (None) 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"nat\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Compile Contract Storage (Record) 1`] = `
"
        (
            template_id (static_id 0 (\\"Compilation.test.ts\\" 125))
            storage (record (\\"Compilation.test.ts\\" 126) (testField1 (literal (nat 1) (\\"Compilation.test.ts\\" 127))) (testField2 (list (\\"Compilation.test.ts\\" 128) (literal (string \\"Hello World\\") (\\"Compilation.test.ts\\" 128)))))
            storage_type (())
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"Compilation.test.ts\\" 125))
        )
        "
`;

exports[`Compile Contract Storage (Record) 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"pair\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"nat\\",
                        \\"annots\\": [
                            \\"%testField1\\"
                        ]
                    },
                    {
                        \\"prim\\": \\"list\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"string\\"
                            }
                        ],
                        \\"annots\\": [
                            \\"%testField2\\"
                        ]
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Compile Contract Storage (Some) 1`] = `
"
        (
            template_id (static_id 0 (\\"Compilation.test.ts\\" 114))
            storage (variant \\"Some\\" (literal (nat 1) (\\"Compilation.test.ts\\" 114)) (\\"Compilation.test.ts\\" 114))
            storage_type (())
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"Compilation.test.ts\\" 114))
        )
        "
`;

exports[`Compile Contract Storage (Some) 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"option\\",
                \\"args\\": [
                    {
                        \\"prim\\": \\"nat\\"
                    }
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Compile Contract Storage (Unit) 1`] = `
"
        (
            template_id (static_id 0 (\\"Compilation.test.ts\\" 108))
            storage (unit)
            storage_type (())
            messages ()
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"Compilation.test.ts\\" 108))
        )
        "
`;

exports[`Compile Contract Storage (Unit) 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Compile Lambdas A 1`] = `"(literal (nat 1) (\\"Compilation.test.ts\\" 46))"`;

exports[`Compile Lambdas A 2`] = `
"{
    \\"micheline\\": \\"1\\",
    \\"json\\": {
        \\"int\\": \\"1\\"
    }
}"
`;

exports[`Compile Lambdas A Lambda that returns "YES" if the argument is greater than or equal to Nat(10), returns "NO" otherwise 1`] = `"(lambda 2 None False \\"lambda_arg\\" (\\"Compilation.test.ts\\" 37) ((ifBlock (ge (lambdaParams 2 \\"lambda_arg\\" (\\"Compilation.test.ts\\" 37) (unknown 0)) (literal (nat 1) (\\"Compilation.test.ts\\" 37)) (\\"Compilation.test.ts\\" 37)) ((result (literal (string \\"YES\\") (\\"Compilation.test.ts\\" 37)) (\\"Compilation.test.ts\\" 37))) (\\"Compilation.test.ts\\" 37)) (elseBlock ((result (literal (string \\"NO\\") (\\"Compilation.test.ts\\" 37)) (\\"Compilation.test.ts\\" 37))))))"`;

exports[`Compile Lambdas A Lambda that returns "YES" if the argument is greater than or equal to Nat(10), returns "NO" otherwise 2`] = `
"{
    \\"micheline\\": \\"{ PUSH nat 1; SWAP; COMPARE; GE; IF { PUSH string \\\\\\"YES\\\\\\" } { PUSH string \\\\\\"NO\\\\\\" } }\\",
    \\"json\\": [
        {
            \\"prim\\": \\"PUSH\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"nat\\"
                },
                {
                    \\"int\\": \\"1\\"
                }
            ]
        },
        {
            \\"prim\\": \\"SWAP\\"
        },
        {
            \\"prim\\": \\"COMPARE\\"
        },
        {
            \\"prim\\": \\"GE\\"
        },
        {
            \\"prim\\": \\"IF\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"string\\"
                            },
                            {
                                \\"string\\": \\"YES\\"
                            }
                        ]
                    }
                ],
                [
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"string\\"
                            },
                            {
                                \\"string\\": \\"NO\\"
                            }
                        ]
                    }
                ]
            ]
        }
    ]
}"
`;

exports[`Compile Lambdas A Lambda that returns the argument 1`] = `"(lambda 1 None False \\"lambda_arg\\" (\\"Compilation.test.ts\\" 28) ((ifBlock (eq (lambdaParams 1 \\"lambda_arg\\" (\\"Compilation.test.ts\\" 28) (unknown 0)) (literal (string \\"TEST\\") (\\"Compilation.test.ts\\" 28)) (\\"Compilation.test.ts\\" 28)) ((result (lambdaParams 1 \\"lambda_arg\\" (\\"Compilation.test.ts\\" 28) (unknown 0)) (\\"Compilation.test.ts\\" 28))) (\\"Compilation.test.ts\\" 28)) (elseBlock ((failwith (lambdaParams 1 \\"lambda_arg\\" (\\"Compilation.test.ts\\" 28) (unknown 0)) (\\"Compilation.test.ts\\" 28))))))"`;

exports[`Compile Lambdas A Lambda that returns the argument 2`] = `
"{
    \\"micheline\\": \\"{ DUP; PUSH string \\\\\\"TEST\\\\\\"; COMPARE; EQ; IF {} { FAILWITH } }\\",
    \\"json\\": [
        {
            \\"prim\\": \\"DUP\\"
        },
        {
            \\"prim\\": \\"PUSH\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"string\\"
                },
                {
                    \\"string\\": \\"TEST\\"
                }
            ]
        },
        {
            \\"prim\\": \\"COMPARE\\"
        },
        {
            \\"prim\\": \\"EQ\\"
        },
        {
            \\"prim\\": \\"IF\\",
            \\"args\\": [
                [],
                [
                    {
                        \\"prim\\": \\"FAILWITH\\"
                    }
                ]
            ]
        }
    ]
}"
`;
